core/src/lib/channel.h

```diff
@@ -0,0 +1,349 @@
+      return intern.data.size() > 0 || intern.in_dead;
+    });
+    if (locked->data.size() == 0) {
+      return std::nullopt;
+    } else {
+      return std::make_optional<handle>(std::move(locked), &out_update);
+    }
+  }
+
+  std::optional<handle> write_lock()
+  {
+    auto locked = shared.lock();
+    locked.wait(out_update, [max_size = max_size](const auto& intern) {
+      return intern.data.size() < max_size || intern.out_dead;
+    });
+    if (locked->in_dead) {
+      // note(ssura): This happening is programmer error.
+      // Maybe we should assert this instead ?
+      Dmsg0(50,
+            "Tried to write to channel that was closed from the write side.\n");
+      return std::nullopt;
+    }
+    if (locked->out_dead) {
+      return std::nullopt;
+    } else {
+      return std::make_optional<handle>(std::move(locked), &in_update);
+    }
+  }
+
+  using try_result
+      = std::variant<handle, failed_to_acquire_lock, channel_closed>;
+
+  try_result try_read_lock()
+  {
+    auto locked = shared.try_lock();
+    if (!locked) { return failed_to_acquire_lock{}; }
+    if (locked.value()->out_dead) {
+      // note(ssura): This happening is programmer error.
+      // Maybe we should assert this instead ?
+      Dmsg0(50,
+            "Tried to read from channel that was closed from the read side.\n");
+      return channel_closed{};
+    }
+    if (locked.value()->data.size() == 0) {
+      if (locked.value()->in_dead) {
+        return channel_closed{};
+      } else {
+        return failed_to_acquire_lock{};
+      }
+    }
+
+    return try_result(std::in_place_type<handle>, std::move(locked).value(),
+                      &out_update);
+  }
+
+  try_result try_write_lock()
+  {
+    auto locked = shared.try_lock();
+    if (!locked) { return failed_to_acquire_lock{}; }
+    if (locked.value()->in_dead) {
+      // note(ssura): This happening is programmer error.
+      // Maybe we should assert this instead ?
+      Dmsg0(50,
+            "Tried to write to channel that was closed from the write side.\n");
+      return channel_closed{};
+    }
+    if (locked.value()->out_dead) { return channel_closed{}; }
+    if (locked.value()->data.size() >= max_size) {
+      return failed_to_acquire_lock{};
+    }
+
+    return try_result(std::in_place_type<handle>, std::move(locked).value(),
+                      &in_update);
+  }
+
+  void close_in()
+  {
+    shared.lock()->in_dead = true;
+    in_update.notify_one();
+  }
+
+  void close_out()
+  {
+    shared.lock()->out_dead = true;
+    out_update.notify_one();
+  }
+};
+
+template <typename T> class in {
+  std::shared_ptr<queue<T>> shared;
+  bool did_close{false};
+
+ public:
+  in(std::shared_ptr<queue<T>> shared) : shared{std::move(shared)} {}
+  in(in&&) = default;
+  in& operator=(in&&) = default;
+  in(const in&) = delete;
+  in& operator=(const in&) = delete;
+
+  template <typename... Args> bool emplace(Args... args)
```

Comment:
when we have `emplace()` maybe we should also have `push_back()`?

Reply:
I tried to keep the interface as small as possible.  Do you know of a situation where you actually want `push_back()` but not `emplace()` ?
If you just pass something to `emplace()` without "moving" it, it should call the copy constructor which should do what `push_back()` does.